move2kube/exercise-3-customizations/exercise-3b-starlark-based-transformers/README.md

```diff
@@ -1 +1,209 @@
+  ```
+  ```yaml
+  apiVersion: move2kube.konveyor.io/v1alpha1
+  kind: Transformer
+  metadata:
+    name: IngressAnnotator
+    labels: 
+      move2kube.konveyor.io/built-in: false
+  spec:
+    class: "Starlark"
+    consumes:
+      KubernetesYamls: 
+        merge: false
+        # Ensures an artifact of this type gets immediately intercepted by this transformer
+        mode: "MandatoryPassThrough" 
+    produces:
+      KubernetesYamls:
+        disabled: false
+    config:
+      starFile: "ingress-annotator.star"
+  ```
+
+The code of the starlark script is shown below. At a high-level, the code requires only the `transform()` function as it acts upon any kubernetes yaml generated within Move2Kube. The `transform()` function loops through every yaml generated for every detected service, checks whether it is an ingress yaml, and if so, adds the annotation. The path mappings are meant to persist these changes.
+```console
+$ cat customizations/add-custom-kubernetes-annotation/ingress-annotator.star
+```
+```python
+def transform(new_artifacts, old_artifacts):
+    pathMappings = []
+    artifacts = []
+
+    for a in new_artifacts:
+        yamlsPath = a["paths"]["KubernetesYamls"][0]
+        serviceName = a["name"]
+        artifacts.append(a)
+
+        fileList = fs.read_dir(yamlsPath)
+        yamlsBasePath = yamlsPath.split("/")[-1]
+        # Create a custom path template for the service, whose values gets filled and can be used in other pathmappings
+        pathTemplateName = serviceName + yamlsBasePath
+        pathTemplateName = pathTemplateName.replace("-", "")
+        tplPathData = {'PathTemplateName': pathTemplateName}
+        pathMappings.append({'type': 'PathTemplate', \
+                            'sourcePath': "{{ OutputRel \"" + yamlsPath + "\" }}", \
+                            'templateConfig': tplPathData})
+        for f in fileList:
+            filePath = fs.path_join(yamlsPath, f)
+            s = fs.read(filePath)
+            yamlData = yaml.loads(s)
+            if yamlData['kind'] != 'Ingress':
+                continue
+            if 'annotations' not in yamlData['metadata']:
+                yamlData['metadata']['annotations'] = {'kubernetes.io/ingress.class': 'haproxy'}
+            else:
+                yamlData['metadata']['annotations']['kubernetes.io/ingress.class'] = 'haproxy'
+            s = yaml.dumps(yamlData)
+            fs.write(filePath, s)
+            pathMappings.append({'type': 'Default', \
+                    'sourcePath': yamlsPath, \
+                    'destinationPath': "{{ ." + pathTemplateName + " }}"})
+
+    return {'pathMappings': pathMappings, 'artifacts': artifacts}
+```
+
+The above steps can be replicated in the UI, by uploading the zip of the custom transformer as an input of type `customization`. You can get the zip of the source and customization by adding a `-z` to the end of the commands used in step 2 and step 4.
```

Comment:
Can you please also add the zipped customization folder, so that can be directly used in the UI. We are already providing the zip of the source folder here https://github.com/konveyor/hackfest/blob/main/move2kube/source/enterprise-app.zip. Also, please update the Move2Kube UI related instruction accordingly.

Reply:
added the customizations.zip

Reply:
I have updated the last line about where to find the `customizations.zip` file